tools/tcpdrop.bt

```diff
@@ -80,7 +73,17 @@ tracepoint:skb:kfree_skb
 
     time("%H:%M:%S ");
     printf("%-8d %-16s ", pid, comm);
-    printf("%39s:%-6d %39s:%-6d %-10s\n", $saddr, $lport, $daddr, $dport, $statestr);
+        
+    if ($inet_family == AF_INET) {
+      $daddr = ntop($sk->__sk_common.skc_daddr);
+      $saddr = ntop($sk->__sk_common.skc_rcv_saddr);
+      printf("%39s:%-6d %39s:%-6d %-10s\n", $saddr, $lport, $daddr, $dport, $statestr);
+    } else {
+      $daddr = ntop($sk->__sk_common.skc_v6_daddr.in6_u.u6_addr8);
+      $saddr = ntop($sk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr8);
+      printf("%39s:%-6d %39s:%-6d %-10s\n", $saddr, $lport, $daddr, $dport, $statestr);
+    }
```

Comment:
While I'm a fan of block scoping in general, the fact that `$daddr` is not accessible after the `if`/`else` block is quite a big backwards compatibility breakage. I can imagine a lot of scripts out there which rely on this functionality.

I may have missed something in the discussions but do we plan to introduce variable declarations so that the only update people need to do is add an explicit declaration before the condition? Or do we even consider declaration hoisting?

Reply:
This was discussed on IRC a bit but I dont think anything came out of it.

I'm not a fan of hoisting as it basically undoes the scoping. You're lifting the var to the outer scope, might as well stick with probe scope. Nested if statements also become confusing

If we do block scoping we also need a way to do upfront declaration. Either the ugly way :

```
$daddr = ntop(dummy)
if ...
  $daddr = ntop(real)
 ```
 
 or by adding the option to define upfront.
 
 This probably shouldn't land until a decision has been made here so good to bring it up

Reply:
I think the general consensus is that we want to introduce variable declarations, which seem to coincide with both the [user-defined functions PR](https://github.com/bpftrace/bpftrace/pull/3068) and this [map declarations RFC](https://github.com/bpftrace/bpftrace/issues/2954). If we land this PR without waiting for variable declarations then to @fbs 's point, users have to use the ugly way in that they need to initialize with a junk or 0 value. 

IMHO, variable declarations don't seem very controversial and are probably not that hard to implement, so I'm fine doing that work either stacked on top of this PR or to follow closely after before we release another bpftrace version with this in it.

Reply:
So, to make things clear, with variable declarations but no hoisting (which is roughly the plan), the above will become
```
let $daddr, $saddr: inet_t;
if ($inet_family == AF_INET) {
  $daddr = ntop($sk->__sk_common.skc_daddr);
  $saddr = ntop($sk->__sk_common.skc_rcv_saddr);
} else {
  $daddr = ntop($sk->__sk_common.skc_v6_daddr.in6_u.u6_addr8);
  $saddr = ntop($sk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr8);
}
printf("%39s:%-6d %39s:%-6d %-10s\n", $saddr, $lport, $daddr, $dport, $statestr);
```
and semantic analyser will check that the value is defined in all branches, otherwise print a warning.

From users perspective, they will just need to add explicit declarations to their scripts. It's still a backwards compat breakage but certainly something we can live with.

Reply:
@viktormalik That's pretty much how I imagine it - maybe without the `let` keyword 😉

Reply:
@fbs @viktormalik I've got a first draft of variable declarations (still needs more tests and a fix that hasn't been merged yet) just to give you an idea of what it looks like. It's pretty simple, which is good:
https://github.com/jordalgo/bpftrace/commit/2bcb96a35287e39fb20e12cfb121a1e334474aab

Reply:
Looks pretty straightforward, nice!

Reply:
IIRC some other MR/proposal added a let syntax. Should we centralize that work in an rfc like issue and block this MR on that?

Reply:
@fbs Are you talking about the map declarations proposal? https://github.com/bpftrace/bpftrace/issues/2954 (cc @JakeHillion)